[SYCL][Graph] Add spec wording for dynamic events (#372)
reble/llvm: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

Reviewer: Think we need to update our definition of the `ext_oneapi_limited_graph` aspect to include that such a device also doesn't include support for dynamic-events

|===

==== Event Terminology  [[event-terminology]]

:events-spec: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:interface.event

For the purposes of clarity when talking about events in this specification we
will split events into two categories:

> - *Limited graph events*: These are events returned from a queue submission

Reviewer 2: How about to rename "limited graph events" to "graph limited events", to say that the events are limited to be valid within a graph scope?

"Limited graph events" gives me the wrong impression that there is another type of "full graph events",

Author: I've updated the usage to "graph-limited events".

which is recorded to a `command_graph`. These events are only valid for defining
dependencies for other nodes inside a `command_graph`. These events cannot be

> waited on or used as dependencies for normal SYCL operations. They also cannot

Reviewer 2: `normal SYCL operation` is used twice in the spec, while `eager SYCL operation` is used 6 times in the spec.  Shall we use one term for it?

Author: I've updated these to be `eager` in all but one case where I am currently working on some other wording changes which are not ready to push.

to define an edge between existing nodes, or using a
`property::node::depends_on` property list when adding a new node to the graph.
Nodes passed to this property may be from the same graph (creating internal
edges) or other graphs (see <<inter-graph-dependencies, this section>> on
creating dependencies between graphs).

Edges can also be created when explicitly adding nodes to the graph through
existing SYCL mechanisms for expressing dependencies. Data dependencies from
accessors to existing nodes in the graph are captured as an edge. 

Using `handler::depends_on()` inside the node's command-group function can also
be used for defining graph edges. However, for an event passed to
`handler::depends_on()` to create an edge, it must be an event returned from a
queue submission captured by a graph (a <<event-terminology, limited graph
event>>). Limited graph events from the same graph will create internal edges,
and those from another graph will create an <<inter-graph-dependencies, external
dependency>>. Passing events from other sources (<<event-terminology, regular
SYCL events>>) will not create edges in the graph, but will create runtime

> dependencies for a graph node on those other events.

Reviewer 2: just an open.

from line 269 to 290, `edge`, `dependency`, `data dependency`, `internal edge`, `external dependency`, `runtime dependency` are used.

my understanding is that: 
`edge`, `internal edge`, `dependency` and `data dependency` are the same.
`external dependency` and `runtime dependency` are almost the same.

And the scope of `dependency` is larger than the scope of `external/runtime dependency`.

just think they are likely to be confused.

Author: I agree this needed tightening up. I've added some wording to the terminology sections for each API to better differentiate "External dependencies" from talking about graph edges and removed usages of "runtime dependency", as well as removed use of "internal edge(s)" as I felt they were superflous.

I think "dependency" is by nature quite context dependent, and "data dependency" I feel is important as is because it is specifically used in terms of SYCL data dependencies from buffers. Each use of "data dependency" is explicitly in the context of buffers/accessors so I believe this is fine as is.

|===

===== Explicit API Example

Simple example that shows using the explicit API to add two nodes to a graph

> with the <<depends-on-property, depends_on property>> used to define

Reviewer 2: ```suggestion
with the <<depends-on-property, node depends_on property>> used to define
```

through newly added interfaces. This is either using the `make_edge()` function
to define an edge between existing nodes, or using a
`property::node::depends_on` property list when adding a new node to the graph.
Nodes passed to this property that are from the same graph will create edges

> within the graph.

Reviewer 2: ```suggestion
within the graph. Events passed to this property that are from the same graph will create edges within the graph.
```

`property::node::depends_on` accepts both node and event

Edges can also be created when explicitly adding nodes to the graph through
existing SYCL mechanisms for expressing dependencies. Data dependencies from
accessors to existing nodes in the graph are captured as an edge. 

Using `handler::depends_on()` inside the node's command-group function can also
be used for defining graph edges. However, for an event passed to
`handler::depends_on()` to create an edge, it must be an event returned from a

> queue submission captured by a graph (a <<event-terminology, graph-limited

Reviewer 2: ```suggestion
queue submission captured by the same graph (a <<event-terminology, graph-limited
```

event>>). 

| External Dependencies | Graph nodes may have dependencies on operations
outside of the graph they belong to. These can be dependencies on nodes from
<<inter-graph-dependencies, another graph>> or on eager SYCL operations. Passing
<<event-terminology, graph-limited events>> from another graph or
<<event-terminology, regular SYCL events>> to `handler::depends_on()`, as well
as passing nodes from another graph to the `property::node::depends_on`
property, will create external dependencies for graph nodes that will be
respected when the graph is executed.

|===

===== Explicit API Example

Simple example that shows using the explicit API to add two nodes to a graph
with the <<depends-on-property, depends_on node property>> used to define

> dependencies between them.

Reviewer 2: ```suggestion
graph edge between them.
```

represent data dependencies between two command groups captured as nodes.
Secondly, by using the `handler::depends_on()` mechanism inside a command group
captured as a node. However, for an event passed to `handler::depends_on()` to
create an edge, it must be an event returned from a queue submission captured by

> the same graph. Passing events from other sources (other graph submissions,

Reviewer 2: maybe we can list all the sources for the event at some place in the spec. and also give a naming to each event type, and then the spec can be described more clear.

regular sycl submission
sycl graph submission
sycl submission during recording
query a given node of an exec graph
...

Author: We previously had such a description in the spec only to help with understanding the differences, but to my mind there are only two relevant types:

- Events which come from a recorded graph submission (previously described as "Graph-limited events"
- All other events should be considered normal SYCL events for all intents and purposes

@gmlueck Do y ou have any objections to bringing back some kind of definition like this in the spec only to help with clarity? It is a bit verbose having to describe the type of event every single time it is mentioned like in the example above.

regular SYCL submissions) will not create edges in the graph, but will create

> runtime dependencies for a graph node on those other events. Thirdly, for a

Reviewer 2: the runtime dependencies also impacts the explicit API with depends_on(EventTN... events), so shall we also add something at line 247?

Secondly, by using the `handler::depends_on()` mechanism inside a command group
captured as a node. However, for an event passed to `handler::depends_on()` to
create an edge, it must be an event returned from a queue
submission captured by the same graph. Passing events from other sources (other
graph submissions, regular SYCL submissions) will not create edges in the graph,
but will create runtime dependencies for a graph node on those other events.

> `handler::depends_on()` can be used to express edges when a user is working with

Reviewer: > `handler::depends_on()` can be used to express edges when a user is working with
USM memory rather than SYCL buffers.

Thinking that we could remove this sentence to keep the length of this paragraph down. It's general SYCL usage advice and doesn't help specify graph behavior.

represent data dependencies between two command groups captured as nodes.
Secondly, by using the `handler::depends_on()` mechanism inside a command group
captured as a node. However, for an event passed to `handler::depends_on()` to
create an edge, it must be an event returned from a queue submission captured to
the same graph (a <<event-terminology, graph-limited event>>). For a graph
recorded with an in-order queue, an edge is added automatically between two
sequential command groups submitted to the in-order queue.

| External Dependencies
| Graph nodes may have dependencies on operations outside of the graph they
belong to. These can be dependencies on nodes from <<inter-graph-dependencies,
another graph>> or on eager SYCL operations. Passing <<event-terminology,
graph-limited events>> from another graph or <<event-terminology, regular SYCL
events>> to `handler::depends_on()` will create external dependencies for
graph nodes that will be respected when the graph is executed.

|===

===== Queue Recording API Example

Simple example that shows using the Queue Recording API to add two nodes to a

> graph with a `sycl::event` used to define the dependency between them.

Reviewer 2: ```suggestion
graph with a `sycl::event` used to define the graph edge between them.
```

[source, c++]
----
namespace sycl_ext = sycl::ext::oneapi::experimental;

sycl_ext::command_graph Graph {SyclContext, SyclDevice};

Graph.begin_recording(SyclQueue);

sycl::event EventA = SyclQueue.submit(
  [&](sycl::handler& CGH){
    CGH.parallel_for(...);
  }
);

SyclQueue.submit(
  [&](sycl::handler& CGH){
    CGH.depends_on(EventA);
    CGH.parallel_for(...);
  }
);

> 

Reviewer 2: Graph.end_recording();

|`aspect::ext_oneapi_graph`
| Indicates that the device supports all the APIs described in this extension.
|`aspect::ext_oneapi_limited_graph`

> a| Indicates that the device supports all the APIs described in this extension

Reviewer: ```suggestion
| Indicates that the device supports all the APIs described in this extension
```

Doesn't seem to affect the rendering, but sneaky `a`

Author: Ah this is actually to indicate that there are things in the table cell that require ASCIIDOC rendering, either the links or the list I can't remember which.

  public:
    template<typename... NodeTN>
    depends_on(NodeTN... nodes);

    depends_on(const event& depEvent);

>     depends_on(const std::vector<event>& depEvents);

Reviewer 3: This seems inconsistent.  The constructor for node dependencies takes a parameter pack while the constructor for events takes a vector.  It seems like you should use the same style in both cases.

};
}
----

The API for explicitly adding nodes to a `command_graph` includes a
`property_list` parameter. This extension defines the `depends_on` property to
be passed here. `depends_on` may be used in two ways:

* Passing nodes from the same `command_graph` which will create dependencies and
graph edges between those nodes and the node being added. 

* Passing SYCL events will create runtime dependencies for execution of the
graph node but will only create edges if those events are associated with other

> nodes in the same graph.

Reviewer 3: This might be a littler clearer:

> * Passing SYCL events.  If an event represents a recorded node in the same graph, then a graph edge is created between this node and the other node.  Otherwise, a runtime dependency is created between this node and the command that is associated with the event.

What happens if the event represents a node recorded into a different graph?

Author: Thanks I think that is clearer.

>What happens if the event represents a node recorded into a different graph?

In that case still only a runtime dependency is created. For easier reasoning about the structure of the graph I think it's important not to call these edges, otherwise this dependency would essentially make them into a single graph.

Edit: Actually perhaps your question is more about what if the event passed is from the recorded queue submission, rather than obtained via `get_event`?

Reviewer 3: > Edit: Actually perhaps your question is more about what if the event passed is from the recorded queue submission, rather than obtained via `get_event`?

Yes, that was my question.

Author: I think right now this would have to still be an error and we should advise to use `get_event()` and `dynamic_event`s to establish these dependencies. This is more explicit which I think is good.

It might be possible to support this in the runtime but there is a very high implementation burden there I think. Neither the graph runtime or UR command-buffers are designed to accomodate that sort of sharing.

Reviewer 3: I agree with this behavior, but the wording should be adjusted to make it clear that this is an error.  Also, what is the behavior if this is violated?  Does `node::add` throw an exception?

Author: Agreed, I've clarified the wording here. Yes `node::add` should throw an exception, but these were not defined so I've added them as well.

  public:
    template<typename... NodeTN>
    depends_on(NodeTN... nodes);

    template<typename... EventTN>
    depends_on(EventTN... events);
};
}
----

The API for explicitly adding nodes to a `command_graph` includes a

> `property_list` parameter. This extension defines the `depends_on` property to

Reviewer: Creating a new thread to branch out of the https://github.com/reble/llvm/pull/372/files#r1647663947 thread, rather than taking over that with a tangent.

> Considering this more, I'm concerned that get_event(node) may not be implementable on a CUDA backend. Since the executable graph has already been finalized then we can't add nodes to it, but to implement this ontop of CUDA-Graph I think we need to use cudaGraphAddEventRecordNode() - which would require some mechanism in the modifiable graph to say that a node is going to produce a sycl event.

I think we need a mechanism for the user to tell the runtime that a node in a modifiable graph will signal out of the graph, so that when we finalize the graph we can create a `cudaGraphAddEventRecordNode` to do that.

My first thought is to add a property that can be set on a node to enable this, and then only nodes with this property can be used in `get_event()`. My other idea was a new node type, but that is effectively an empty node with this property set. Open to other ideas, or discussion on whether the current design is indeed unsuited to a CUDA backend.

Reviewer: Actually for L0 we could still use this info for more precise `zeCommandListGetNextCommandIdExp` usage so that we are only setting the `ZE_MUTABLE_COMMAND_EXP_FLAG_SIGNAL_EVENT` and `ZE_MUTABLE_COMMAND_EXP_FLAG_WAIT_EVENTS` on commands which the users has stated will be used with dynamic events.

Author: After some internal discussion on this topic we settled on the idea of:

- A node property for the explicit API which marks a node as requiring a backend event which can be obtained via `get_event()`
- A graph property usable with `command_graph::begin_recording()` which opts-in to this for all recorded nodes from a queue/queues until `end_recording()` is called.

This allows fine grained opt-in to these required events on both APIs.

Reviewer 3: This sounds reasonable, but I wouldn't call this a "backend event".  It's just an `event`.  Calling it "backend event" makes it sound like it returns a native event from the backend, which is not the case.

Author: Good point! I will avoid such phrasing in the wording here.

be passed here. `depends_on` may be used in two ways:

* Passing nodes from the same `command_graph` which will create dependencies and
graph edges between those nodes and the node being added. 

* Passing SYCL events, including <<dynamic-event, Dynamic Events>>. If an event

> represents a recorded node in the same graph, then a graph edge is created

Reviewer 2: ```suggestion
represents a node in the same graph, then a graph edge is created
```
"recorded" may give a wrong impression that only for the recorded graph, not the graph created with explicit APIs.

Author: There currently isn't a way to get an event like this for nodes added explicitly, though you can go the other way and get a `node` for a given event from a recorded queue submission so this is probably a gap that should be addressed.

Reviewer 3: Doesn't `command_graph::get_event` provide this ability?

Author: > Doesn't command_graph::get_event provide this ability?

Since `get_event()` can only be called on executable graphs and returns an event specifically representing an execution point in a single graph execution (and can be used largely as normal SYCL events), they are conceptually quite different from the events returned from recorded queue submissions. 

Those have no backend events associated with them, have many restrictions on where they can be used and are only for use in creating dependencies between graph nodes.

between this node and the other node. Passing an event from a recorded
submission to another graph is an error (see <<event-limitations, Event

> Limitations>> for more information). For events obtained from another graph via

Reviewer 2: ```suggestion
Limitations>> for more information). For `dynamic_event` and `event`s obtained from another graph via
```

Author: I've modified the wording to this effect but slightly changed, since the suggested wording implies dynamic events are obtained from graphs.

  public:
    template<typename... NodeTN>
    depends_on(NodeTN... nodes);

    depends_on(const event& ev);

>     depends_on(const std::vector<event>& events);

Reviewer: nit - `handler::depends_on` names these parameters as `depEvent` and `depEvents` -  suggest staying consistent with rather than `ev` & `events`

};
}
----

The API for explicitly adding nodes to a `command_graph` includes a
`property_list` parameter. This extension defines the `depends_on` property to
be passed here. `depends_on` may be used in two ways:

> * Passing nodes from the same command_graph which will create dependencies and

Reviewer: ```suggestion
* Passing nodes from the same `command_graph `which will create dependencies and
```

    void update(node& node);
    void update(const std::vector<node>& nodes);

>     event get_event(const node& n);

Reviewer: Preference for not using single letter variable names
```suggestion
    event get_event(const node& node);
```

Author: The issue with `node` is it collides with the actual class name which is not great practice I think, having trouble thinking of a prefix for `node` that makes sense in this context though. Best I can think of is `targetNode` but IMO that doesn't express it very well either.

Reviewer: I was mostly thinking of this from a consistency perspective, where we have `void update(node& node);` but in `graph.hpp` we define it as `void update(node& Node);` .

Author: That is fair, I've made that change to `node` then.

  public:
    template<typename... NodeTN>
    depends_on(NodeTN... nodes);

    template<typename... EventTN>
    depends_on(EventTN... events);
};
}
----

The API for explicitly adding nodes to a `command_graph` includes a
`property_list` parameter. This extension defines the `depends_on` property to
be passed here. `depends_on` may be used in two ways:

* Passing nodes from the same `command_graph` which will create dependencies and
graph edges between those nodes and the node being added. 

> * Passing SYCL events, including <<dynamic-event, Dynamic Events>>. If an event

Reviewer 2: ```suggestion
* Passing SYCL events, including <<dynamic-events, Dynamic Events>>. If an event
```

> is a <<event-terminology, graph-limited event>>, then a graph edge is created

Reviewer 2: ```suggestion
is a <<event-terminology, graph-limited event>> from the same `command_graph`, then a graph edge is created
```

graph and will not affect previous submissions of the same graph. The user is
not required to wait on any previous submissions of a graph before updating it.

The aspects of a kernel execution node that can be configured during update are:

* Parameters to the kernel.
* Execution ND-Range of the kernel.

All node types may have the following aspects configured during update:

> * Dependent events which were specified using <<dyanmic-events, Dynamic

Reviewer 4: ```suggestion
* Dependent events which were specified using <<dynamic-events, Dynamic
```

use the buffer as a parameter. Then a single `dynamic_parameter::update()` call
will maintain the graphs data dependencies.

===== Node Event Dependency Update

Event dependencies for nodes can be updated using <<dynamic-events, Dynamic

> Events>> in a similar usage to Dynamic Parameters.

Reviewer: "Dynamic Parameters" could also be a link

* Throws synchronously with error code `invalid` if a queue is recording
  commands to the graph.

* Throws synchronously with error code `invalid` if the graph wasn't created with
  the `property::graph::assume_buffer_outlives_graph` property and this command
  uses a buffer. See the
  <<assume-buffer-outlives-graph-property, Assume-Buffer-Outlives-Graph>>
  property for more information.

* Throws with error code `invalid` if the type of the command-group is not a
  kernel execution and a `dynamic_parameter` was registered inside `cgf`.

> * Throws synchronously with error code `invalid` if an `event` dependency is

Reviewer 5: Maybe I misunderstood something, but isn't this inconsistent with the line 283:

```
However, for an event passed to `handler::depends_on()` to
create an edge, it must be an event returned from a queue submission captured by
the same graph. Passing events from other sources (other graph submissions,
egular SYCL submissions) will not create edges in the graph, but will create
runtime dependencies for a graph node on those other events.
```

Doesn't this imply that passing an event from a recorded graph should be acceptable ? Is this a difference between:

event: end of graph replay
event: some event that is in the middle of a recorded graph

?

I think that would make sense, but I'm not sure if this was explicitly noted in any of the places, unless I overlooked something (which might have happened, ofc).

Reviewer 5: EDIT: I've checked the code sample and I don't think I understand this

```
auto nodeB = graphB.add([&](handler& CGH){
  CGH.depends_on(dynEvent);
```

here, we are depending on event dependency from another recorded graph; is this ok because the event is wrapped into "dynamic" event or, maybe, because the first graph was an explicit graph, with node added via `graphA.add(...);` ?

Author: I agree the wording around these types of events is a little confusing and I will try to push some changes soon to clarify, but perhaps I can clarify here as well.

>  Is this a difference between:
>
> event: end of graph replay
> event: some event that is in the middle of a recorded graph

Yes there is a difference here. Events returned from a recorded queue submission to a graph, as in:

```c++
Graph.begin_recording(Queue);
auto EventA = Queue.submit(...);

Queue.submit((sycl::handler& CGH){
// Valid because it came from a previous submission to the same graph
CGH.depends_on(EventA);
...
});
Graph.end_recording();
```
These events have no associated device event (because nothing is created on the device until `finalize()` is called) and so we only use them for establishing internal edges within a graph. So they can be used with `handler::depends_on()` inside another submission to the same graph but not waited on or used as dependencies for normal SYCL submissions or a separate graph's nodes/recorded submissions.

Events returned from a submission of a graph, i.e. `auto ExecEvent = queue::ext_oneapi_graph(execGraph);` are normal SYCL events like those returned from other normal queue submissions, and that event represents the completion of the entire graph.

The second example involving dynamic events is a slightly different case. When a graph in the modifiable state is finalized into an executable graph, it is now possible to obtain SYCL events backed by device events using the `command_graph<state::executable>::get_event(Node)` API. This is because `finalize()` is the step that actually creates the device objects for execution, and so it is now possible to look up the device event associated with a particular node in an executable graph, and return it as a SYCL event to the user.  This resulting SYCL event can be used as any normal SYCL event can.

Because it is a normal SYCL event it can be used as a dependency for nodes in another graph (as seen in the example), this creates an "external" event dependency not an internal edge in the graph. External event dependencies can also be created from any other normal SYCL events (such as those returned from normal queue submissions).

To complicate this slightly I am going to be making some changes to the wording around the use of `get_event()` described here based on discussion [here](https://github.com/reble/llvm/pull/372#discussion_r1700591179) which may change slightly how this works for establishing dependencies between different graphs. Allowing the use of the events I first described (returned from recorded queue submissions, with no device event backing) to establish dependencies between graphs directly has also been discussed [here](https://github.com/reble/llvm/pull/372#discussion_r1617262562), but that would likely be done as a future update to this part of the spec not as part of this initial work.

Sorry for the very verbose reply but hopefully that clarifies things as they stand right now! Please let me know if you have further comments or questions though.

----
|Returns a list of all nodes in the graph which have no dependencies.

|
[source,c++]
----
event get_event(const node& n);
----
|Returns a SYCL event which represents the completion of node `n` which is valid
only for the next execution of the graph. This event can be used as a dependency
in the same way as normal SYCL events.

> 

Reviewer: I'd expect some more sections here, e.g.:
```
Constraints:
* This member function is only available when the `command_graph` state is
  `graph_state::executable`.

Parameters:
* `node` - foo

Exceptions:
* Throws with error code `invalid` if `node` is not part of the
  graph.

```

Author: Done!

----
|Returns a list of all nodes in the graph which have no dependencies.

|
[source,c++]
----
event get_event(const node& node);
----
|Returns a SYCL event which represents the completion of node `node` which is

> valid only for the next execution of the graph. This event can be used as a

Reviewer 2: I'm curious why the event is valid only for the `next` graph execution.  In graph.finalize(), the event is tied with a native event, and I think (at lease for most cases), the event is fixed after graph.finalize(). 

Or the actual purpose of this function is to get a chance to reset the fixed event?

Author: SYCL events don't have any concept of being reset, and if we introduced that it becomes unclear when events are reset if it happens automatically every time you submit a graph for execution.

This way it is explicit what an event represents. It is not necessarily true that the backend event will be reset by this function call, but that is the general idea.

Reviewer 2: > It is not necessarily true that the backend event will be reset by this function call, but that is the general idea.

Does this mean that "the backend event will be reset by this function call" is the general idea in practice, but it is not a must?  (From spec perspective, it does not define this function's implementation.)

Suppose one implementation chooses to reset the backend event for this function, and the event is not reused by other places, based on it, just a "thought experiment",  the code below:
```
dynEvent.update(execGraphA.get_event(nodeA));
execGraphB.update(nodeB);
```
can be simplified with below code:
```
execGraphA.get_event(nodeA);
```
Is it correct for this special implementation? The reason is that the backend event used by execGraphA does not change (just the event status is changed), and so the two "update" functions can be removed for a certain implementation.

Author: I think my previous reply was poorly worded, I did not mean to imply that that is how the implementation will behave, and I don't think an implementation would or should allow your simplified code.

This is basically the same idea being discussed [here](https://github.com/reble/llvm/pull/372#discussion_r1617576585) if I'm not mistaken, and I would reiterate that is something we can explore in the future but not as part of this initial work.

Reviewer 2: what's the expected/typical implementation of this function to make the returned event only valid for the next execution of the graph, thanks.

Author: I'm afraid I don't really have an answer for that immediately, but I think you are right that this needs some further consideration so we are looking into it.

Reviewer: Considering this more, I'm concerned that `get_event(node)` may not be implementable on a CUDA backend. Since the executable graph has already been finalized then we can't add nodes to it, but to implement this ontop of CUDA-Graph I think we need to use `cudaGraphAddEventRecordNode()` - which would require some mechanism in the modifiable graph to say that a node is going to produce a sycl event.

EDIT: I've moved this comment to its own thread here https://github.com/reble/llvm/pull/372#discussion_r1648714832 to avoid branching the discussion from performant usage on backend that can reset events.

Reviewer: Wondering if defining the semantics of dynamic_event like a counter-based event would help give a more optimized flow for the node-to-node only use case. 
* If a `dynamic_event` is created from a vanilla sycl queue submission. Counter only incremented once when sycl event backing it is complete.
* If a `dynamic_event` is created for a node in a graph, then counter incremented on completion of each execution of the graph.

* When waiting on a `dynamic_event` in a vanilla queue submission, waiting on value to increment from current value.

A user wouldn't need to manually call any reset methods or the update() calls currently needed for the node-to-node between graph case. 

Edit: I've actually realized that to implement `get_event(node)`  we may not need to actually update the signal event in a L0 command, but always use the underlying L0 event that already exist for sync-points (or create one if need be) wrapped in different ur events events.

This L0 event gets reset between executions anyway, So we may actually get this kind of counter based
dynamic_event semantics by default without having to go out our way to achieve it.

> dependency in the same way as normal SYCL events.

Reviewer: Wondering if we should have the same wording as `update()` here around blocking behaviour

> The implementation may perform a blocking wait during this call on any in-flight executions of that same graph if required by the backend.

Since until we've done more implementation work I don't think we'll be able to say for sure whether this can be non-blocking.

Author: Following internal discussion:

- `get_event()` would be forced to sync if updating the event in the backend structures is required to provide a unique event per execution.
- The following example illustrates a more general problem with the event returned from `get_event()` being unique per execution:
-  ```c++
   sycl::event EventA = ExecGraph.get_event(SomeNode);
				
   Queue.ext_oneapi_graph(ExecGraph);
   sycl::event EventB = ExecGraph.get_event(SomeNode);
   // We haven't waited on eventA so it is reset and now valid only for the second execution
   // and is the same as eventB?
   Queue.ext_oneapi_graph(ExecGraph);
				
   EventA.wait();
   EventB.wait();
```
- Implementing this so that events are unique would require a host-sync when `EventB` is obtained to update the backend structures with a new event. This would render `EventA` complete anyway.
- In either case effectively this means that events are only valid for the most recent execution of the graph, so we should make this the official wording. If a user requires to wait specifically on each event triggering they must enforce the ordering somehow, or use separate executable graphs to avoid event reset.

                   event depEvent);
  event ext_oneapi_graph(command_graph<graph_state::executable>& graph,
                   const std::vector<event>& depEvents);

>   // Overload of ext_oneapi_barrier which takes dynamic events

Reviewer 6: ```suggestion
  // Overload of ext_oneapi_barrier which takes a list of dynamic events
```
For consistency?

associated with the graph node resulting from this command-group submission is
different from the one with which the dynamic_parameter was created.

|===

=== Dynamic Events [[dynamic-events]]

[source,c++]
----

> namespace ext::oneapi::experimental{
>   class dynamic_event{

Reviewer: consistency nitpick
```suggestion
namespace ext::oneapi::experimental {
  class dynamic_event {
```

    dynamic_event();
    dynamic_event(const event& ev);

    void update(const event& ev);
  };
}
----

Dynamic events represent SYCL events from outside of a given `command_graph`
(either obtained from normal SYCL operations or from another `command_graph`)

> that nodes in that graph may depend on.  Dynamic events also allow for these

Reviewer: Whitespace nitpick but I'd actually maybe rephrase the second sentence as "The `dynamic_event` object enables these dependent events to be updated between graph executions."
```suggestion
that nodes in that graph may depend on. Dynamic events also allow for these
```

  CGH.parallel_for(...);
});
----
Dynamic events created with an event from a `command_graph` cannot then be
associated with other nodes in that same graph as this would change the shape of
the graph. Attempting to call `handler::depends_on()` with such a
`dynamic_event` in that situation will result in an error.

Dynamic events can be created with no event but must be updated with a valid
event before any executable graph which depends on that event is executed.
Failing to do so will result in an error.

Table {counter: tableNumber}. Member functions of the `dynamic_event` class.
[cols="2a,a"]
|===
|Member function|Description

|
[source,c++]
----
dynamic_event();
----

| Constructs a default `dynamic_event` which is not associated with any SYCL event.

|
[source,c++]
----
dynamic_event(const event& ev);
----

| Constructs a `dynamic_event` which is associated with the SYCL event `ev`.

|
[source,c++]
----
void update(const event& ev);
----

| Updates the SYCL event associated with `dynamic_event`. This update will be
reflected immediately in the associated modifiable graph nodes. An executable
graph can then be updated to reflect these new event dependencies using

> <<executable-graph-update, Executable Graph Update>>.

Reviewer: Not sure if that Executable Graph Update intro text needs updated - it's says "The only type of nodes that are currently supported for updating in a graph are kernel execution nodes." Which I don't think is a restriction for this feature.

Author: I've updated the wording there to specify that all node types can have dynamic events updated.

associated with the graph node resulting from this command-group submission is
different from the one with which the dynamic_parameter was created.

|===

=== Dynamic Events [[dynamic-events]]

[source,c++]
----
namespace ext::oneapi::experimental {
  class dynamic_event {
    dynamic_event();
    dynamic_event(const event& ev);

    void update(const event& ev);
  };
}
----

Dynamic events represent SYCL events from outside of a given `command_graph`
(either obtained from normal SYCL operations or from another `command_graph`)
that nodes in that graph may depend on. The `dynamic_event` object enables these
dependent events to be updated between graph executions.

Dynamic events can be used to add dependencies to a graph node in the same way
that regular SYCL events can, by passing them as parameters to
`handler::depends_on()` inside the CGF which represents the node.

> 

Reviewer: For dynamic parameters we have the statement:

> The dynamic_parameter class provides the [common reference semantics](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:reference-semantics).

Should we have something similar here.

Author: Done!

event OutsideEvent = queue.submit(...);

// Create a dynamic event to wrap that event
ext::oneapi::experimental::dynamic_event DynEvent {OutsideEvent};

// Add a graph node which depends on that dynamic event
Graph.add([&](handler& CGH){
  CGH.depends_on(DynEvent);
  CGH.parallel_for(...);
});
----
Dynamic events created with an event from a `command_graph` cannot then be
associated with other nodes in that same graph as this would change the shape of
the graph. Attempting to call `handler::depends_on()` with such a
`dynamic_event` in that situation will result in an error.

Dynamic events can be created with no event but must be updated with a valid
event before any executable graph which depends on that event is executed.
Failing to do so will result in an error.

Table {counter: tableNumber}. Member functions of the `dynamic_event` class.
[cols="2a,a"]
|===
|Member function|Description

|
[source,c++]
----
dynamic_event();
----

| Constructs a default `dynamic_event` which is not associated with any SYCL event.

|
[source,c++]
----
dynamic_event(const event& ev);
----

| Constructs a `dynamic_event` which is associated with the SYCL event `ev`.

|
[source,c++]
----

> void update(const event& ev);

Reviewer: Such a nitpick, but to be consistent with other entry-point descriptions we could have:

```
Parameters:
* `ev` - foo
```

associated with the graph node resulting from this command-group submission is
different from the one with which the dynamic_parameter was created.

|===

=== Dynamic Events [[dynamic-events]]

[source,c++]
----
namespace ext::oneapi::experimental {
  class dynamic_event {
    dynamic_event();
    dynamic_event(const event& syclEvent);

    void update(const event& syclEvent);
  };
}
----

Dynamic events represent SYCL events from outside of a given `command_graph`
(either obtained from normal SYCL operations or from another `command_graph`)
that nodes in that graph may depend on. The `dynamic_event` object enables these
dependent events to be updated between graph executions.

Dynamic events can be used to add dependencies to a graph node in the same way
that regular SYCL events can, by passing them as parameters to
`handler::depends_on()` inside the CGF which represents the node.

[source,c++]
----
// Obtain an event from a normal queue submission
event OutsideEvent = queue.submit(...);

// Create a dynamic event to wrap that event

> ext::oneapi::experimental::dynamic_event DynEvent {OutsideEvent};

Reviewer 2: OutsideEvent will be signaled once the workload submitted at line 1647 `queue.submit(...)` is finished.  Will OutsideEvent be always kept on (and so that line 1654 `CGH.depends_on(DynEvent)` is always satisfied after the first run)?

Author: Yes it would be always satisfied after the first run in that case.

// Add a graph node which depends on that dynamic event
Graph.add([&](handler& CGH){
  CGH.depends_on(DynEvent);
  CGH.parallel_for(...);
});
----
Dynamic events created with an event from a `command_graph` cannot then be

> associated with other nodes in that same graph as this would change the shape of

Reviewer 2: "change the shape of the graph", does this mean that an extra edge is added?  (it is not the purpose of dynamic event.)

Author: Perhaps it would be more accurate to say, "_could_ be used to change the shape of the graph".

If you have nodes A, B and C in a graph and C depends on B through a dynamic event, you could then update the event so instead C depends on A. This would change the structure/topology/flow of the graph, and this is not allowed.

Reviewer 2: i see, thanks. We may replace "shape" with "topology".

Author: Thanks I've made that change!

associated with the graph node resulting from this command-group submission is
different from the one with which the dynamic_parameter was created.

|===

=== Dynamic Events [[dynamic-events]]

[source,c++]
----
namespace ext::oneapi::experimental {
  class dynamic_event {
    dynamic_event();
    dynamic_event(const event& syclEvent);

    void update(const event& syclEvent);
  };
}
----

Dynamic events represent SYCL events from outside of a given `command_graph`
(either obtained from normal SYCL operations or from another `command_graph`)

> that nodes in that graph may depend on. The `dynamic_event` object enables these

Reviewer 2: > that nodes in that graph may depend on. 

not quite understand the meaning with the two "that".

Author: I've reworded this section a little so hopefully that makes more sense now.

dependent events to be updated between graph executions.

Dynamic events can be used to add dependencies to a graph node in the same way
that regular SYCL events can, by passing them as parameters to
`handler::depends_on()` inside the CGF which represents the node.

[source,c++]
----
// Obtain an event from a normal queue submission
event OutsideEvent = queue.submit(...);

// Create a dynamic event to wrap that event
ext::oneapi::experimental::dynamic_event DynEvent {OutsideEvent};

// Add a graph node which depends on that dynamic event
Graph.add([&](handler& CGH){
  CGH.depends_on(DynEvent);
  CGH.parallel_for(...);
});

> ----

Reviewer 2: can we next update DynEvent in this case? for example,

// Obtain another event from a normal queue submission
event e2 = queue.submit(...);
// or
event e3= anotherGraph.get_event(oneNode);

// and then
dynEvent.update(e2 or e3);

execGraph = Graph.finalize({sycl_ext::property::graph::updatable});
execGraph.update(node_added_at_line_1656);

Author: Yes, although in this case you would not need to call update to have the update to `dynEvent` reflected in `execGraph`, since the update happened before the graph was finalized.

In the same way as with `dynamic_parameter`s, updates to a `dynamic_event` are immediately reflected in the node and by extension the modifiable graph that node is part of.

Reviewer 2: I see, and `execGraph.update(node_added_at_line_1656);` is a must with below code?

// Obtain another event from a normal queue submission
event e2 = queue.submit(...);
// or
event e3= anotherGraph.get_event(oneNode);

execGraph = Graph.finalize({sycl_ext::property::graph::updatable});

// and then
dynEvent.update(e2 or e3);

execGraph.update(node_added_at_line_1656);

Author: Yes

Dynamic events can be created with no event but must be updated with a valid
event before any executable graph which depends on that event is executed.
Failing to do so will result in an error.

The `dynamic_event` class provides the {crs}[common reference semantics].

Table {counter: tableNumber}. Member functions of the `dynamic_event` class.
[cols="2a,a"]
|===
|Member function|Description

|
[source,c++]
----
dynamic_event();
----

| Constructs a default `dynamic_event` which is not associated with any SYCL event.

|
[source,c++]
----
dynamic_event(const event& syclEvent);
----

| Constructs a `dynamic_event` which is associated with the SYCL event `syclEvent`.

Parameters:

* `syclEvent` - The SYCL event to associate this `dynamic_event` with.

|
[source,c++]
----
void update(const event& syclEvent);
----

| Updates the SYCL event associated with `dynamic_event`. This update will be
reflected immediately in the associated modifiable graph nodes. An executable
graph can then be updated to reflect these new event dependencies using
<<executable-graph-update, Executable Graph Update>>.

Parameters:

* `syclEvent` - The new SYCL event to update this `dynamic_event` with.

> 

Reviewer 2: may add Exceptions:

Throws with error code invalid if `syclEvent` is associated with a node in a graph, and it is the same graph that this `dynamic_event` is associated since it will change the topo of the graph.

Author: Added an exception for this, thanks!

associated with the graph node resulting from this command-group submission is
different from the one with which the dynamic_parameter was created.

|
[source,c++]
----
void handler::ext_oneapi_barrier(const std::vector<dynamic_event>& waitList);
----

|Overload of `ext_oneapi_barrier` that takes a list of dynamic events to wait
on.

This function has the same semantics as `ext_oneapi_barrier(const

> std::vector<event>&)`.

Reviewer 3: Do we need the same exceptions you list below for `depends_on`?

Author: Ah I missed those, thanks. Added them now.

|
[source,c++]
----
void handler::depends_on(dynamic_event depEvent);
----

|Overload of `depends_on()` which takes a dynamic event dependency.

This function has the same semantics as `depends_on(event)`.

Parameters:

> * `depEvents` - Dynamic event dependency.

Reviewer: ```suggestion
* `depEvent` - Dynamic event dependency.
```

Exceptions:

* Throws synchronously with error code `invalid` if this function is called from

> a normal SYCL command-group submission.

Reviewer 3: I think we should clarify what "normal" SYCL command-group submission means.  I assume you mean a command submission via `queue::submit`, correct?  What if the queue is in record mode, would that be allowed?

Author: > I assume you mean a command submission via queue::submit, correct?

Yes the intent here is that using dynamic events outside of graphs is not allowed. Would a "regular" queue submission be more acceptable language? I do feel like we need some shorthand way to say this since it comes up multiple times.

Reviewer 3: Reviving an idea that we discussed a while ago ... We could design the API to enforce these restrictions through the type system.  To do this, we'd change `command_graph::add` so that the `cgf` callback does *not* take a `sycl::handler`.  Instead, we'd pass some different object to the callback like `sycl::ext::oneapi::experimental::add_handler`.  The `add_handler` would support most of the same member functions as `handler`, omitting those that cannot be used when constructing a graph.  Member functions that are specific to graphs (like this overload of `ext_oneapi_barrier` taking a vector of `dynamic_event`) would be added only to `add_handler`.  Thus, it would be impossible to call these functions from a "normal" SYCL command-group function.

Author: I'm not very keen on this idea because it excludes these features from recording APIs and creates unituitive differences between command groups in either API.

An alternative could be that we allow the `dynamic_event` overloads of these functions in normal SYCL code and just have it use whatever the current event value of the `dynamic_event` is.

Reviewer 3: > I'm not very keen on this idea because it excludes these features from recording APIs and creates unituitive differences between command groups in either API.

Do you really think people will use recording mode and the explicit APIs at the same time like this?  It seems like users who are willing to change their code would just use the explicit API.  Those users who don't want to change their code will use the recording mode.

> An alternative could be that we allow the `dynamic_event` overloads of these functions in normal SYCL code and just have it use whatever the current event value of the `dynamic_event` is.

This wouldn't really solve the problem because a `dynamic_event` can be constructed with no underlying SYCL event.  So, you'd still need special checking in that case.

Author: > Do you really think people will use recording mode and the explicit APIs at the same time like this?

I don't think simultaneous use of the APIs comes into this, and ignores the main point which is creating a feature disparity between the APIs is very undesirable. The fact that it is unintuitive and also creates additional spec bulk is secondary.

> Those users who don't want to change their code will use the recording mode.

I also don't think you can put API preference into such strict terms as this. Recording style APIs are very popular in existing applications using other existing graph APIs, likely because it involves very little change to the way you write your general SYCL/CUDA/whatever code. Using more advanced features like this will likely always require some amount of modification, and that does not take away from the simple use case where you can wrap your code in a graph without modification, but it also doesn't mean that usage of the recording APIs should be limited to that most simple of use cases only.

> This wouldn't really solve the problem because a dynamic_event can be constructed with no underlying SYCL event. So, you'd still need special checking in that case.

Attempting to use a `dynamic_event` in a graph that has no underlying event is [already an error](https://github.com/reble/llvm/blob/9fe7edfa9c1cb6cd6fbc63c6b62a8f24d8339cbf/sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc?plain=1#L1778) (though that wording is missing from the definitions of the new overloads taking dynamic events, I'll fix that) so this would be an error in both cases if the wording is changed to also include this use.

Failing to do so will result in an error.

The `dynamic_event` class provides the {crs}[common reference semantics].

Table {counter: tableNumber}. Member functions of the `dynamic_event` class.
[cols="2a,a"]
|===
|Member function|Description

|
[source,c++]
----
dynamic_event();
----

| Constructs a default `dynamic_event` which is not associated with any SYCL event.

|
[source,c++]
----
dynamic_event(const event& syclEvent);
----

| Constructs a `dynamic_event` which is associated with the SYCL event `syclEvent`.

Parameters:

* `syclEvent` - The SYCL event to associate this `dynamic_event` with.

|
[source,c++]
----
void update(const event& syclEvent);
----

| Updates the SYCL event associated with `dynamic_event`. This update will be
reflected immediately in the associated modifiable graph nodes. An executable
graph can then be updated to reflect these new event dependencies using
<<executable-graph-update, Executable Graph Update>>.

Parameters:

* `syclEvent` - The new SYCL event to update this `dynamic_event` with.

Exceptions:

* Throws synchronously with error code `invalid` if `syclEvent` is an event

> obtained from a queue submission which was recorded to the same graph any of the

Reviewer 3: This is another use of the word "recorded" that could be confusing.  I assume you aren't talking about a queue that's in record mode here.  I'd suggest searching for the word "record" in this document and using a different word unless you really mean record mode.

Author: In this case I do mean a queue in the recording mode, it is tricky to describe these events since they have no type or state to distinguish them. The intention here is that you cannot pass an event associated with a node from the same graph, that being a node that was obtained from a queue submission which was recorded to a graph.

Reviewer 3: You could clarify this in the spec by using a consistent term to refer to these events, and put that term in italics.  Can we refer to these events as a _recorded edge_?  Then, you could rewrite this description like:

> Throws synchronously with error code `invalid` if `syclEvent` is a _recorded edge_.

Do we really need the rest of the description about being from the same graph?  Wouldn't it be a problem also if the event was a _recorded edge_ from a different graph?

Author: I don't think we can equate an event with a graph edge conceptually, it would have to be a node. Also as discucsed in [this comment](https://github.com/reble/llvm/pull/372#discussion_r1642924090) there is an API gap which should be filled for interop between the explicit and recording APIs to allow getting an event from a node object, so this wording should maybe be something like "if `syclEvent` represents a modifiable graph node".

> Do we really need the rest of the description about being from the same graph? Wouldn't it be a problem also if the event was a recorded edge from a different graph?

Right now yes they should both be disallowed, I'll update the wording here. As discussed in [this comment](https://github.com/reble/llvm/pull/372#discussion_r1617576585) we may want to revisit that in the future to allow events from other graphs to avoid the need for manual updating of these inter-graph dependencies by the user.

Reviewer 3: > so this wording should maybe be something like "if `syclEvent` represents a modifiable graph node".

This doesn't sound right to me, or maybe I don't understand the restriction.  I thought the restriction was only for commands that were added via a queue in record mode.  Doesn't the term "modifiable graph node" apply to nodes added through the explicit API?  Maybe you could write the restriction like this:

> Throws synchronously with error code `invalid` if `syclEvent` represents a _recorded command_.

However, I still think it would be better to define a term that applies to the event itself, rather than the command that it represents.  Therefore, maybe:

> Throws synchronously with error code `invalid` if `syclEvent` is a _recorded event_.

Separately:

> As discussed in https://github.com/reble/llvm/pull/372#discussion_r1617576585 we may want to revisit that in the future to allow events from other graphs to avoid the need for manual updating of these inter-graph dependencies by the user.

This sounds like it re-opens the discussion about "resetting" events.  The problem before was that we did not have a precise definition for when an event is reset.  I guess this is a discussion for another day, though.

Author: > This doesn't sound right to me, or maybe I don't understand the restriction.

The restriction is in place because these events returned from:

- A recorded queue submission or
- A hypothetical call to something like `graph.get_event_from_node(nodeObject);`

are not associated with backend events and are only used for dependency tracking within a graph. So events of this type from another graph are not allowed because there is nothing to create an actual backend dependency on and events from the same graph are not allowed because that is creating an edge within the graph (which cannot be updated).

Getting an event from a node is a gap in the interop functionality between our APIs which I do think is important to support, so I feel like "recorded" in this case is too API specific. Perhaps a "modifable-graph event"? To distinguish it from events obtained from an executable graph via `get_event()` which represent an actual execution.

> This sounds like it re-opens the discussion about "resetting" events.

I don't necessarily think it does since the idea discussed in that comment pertains to using the events detailed above to create these dependencies between graphs, and these events cannot be waited on by the user, so once the graphs are built these dependencies are fixed. Any resetting that occurs is purely an implementation detail, similar to what we already do with backend events in command-buffers right now, just extending that to have dependencies between command-buffers.

Reviewer 3: > The restriction is in place because these events returned from:
> 
> * A hypothetical call to something like `graph.get_event_from_node(nodeObject);`
> 
> are not associated with backend events and are only used for dependency tracking within a graph

There is no function named `get_event_from_node`.  I assume you mean `command_graph::get_event`?  But the intended use of `get_event` is specifically to use it with `dynamic_event::update`.  This is even illustrated in the example below "Dynamic Event Update".

Author: >There is no function named get_event_from_node

Yes there is no function named `get_event_from_node`, I was referencing it as a hypothetical based on the discussion from [this comment](https://github.com/reble/llvm/pull/372#discussion_r1642924090) where I mentioned this is a gap in the interop capabilities that should be addressed. However, as that discussion is still ongoing I have not made that change in this PR yet.

----
void handler::depends_on(dynamic_event depEvent);
----

|Overload of `depends_on()` which takes a dynamic event dependency.

This function has the same semantics as `depends_on(event)`.

Parameters:

* `depEvent` - Dynamic event dependency.

Exceptions:

* Throws synchronously with error code `invalid` if this function is called from
a normal SYCL command-group submission.

* Throws synchronously with error code `invalid` if the `sycl::event` associated
with `depEvent` came from the same graph that the graph node resulting from this
command-group submission is associated with.

|
[source,c++]
----
void handler::depends_on(const std::vector<dynamic_event>& depEvents);
----

|Overload of `depends_on()` which takes a list of dynamic event dependencies.

This function has the same semantics as `depends_on(const std::vector<event>&)`.

Parameters:

* `depEvents` - List of dynamic event dependencies.

Exceptions:

* Throws synchronously with error code `invalid` if this function is called from
a normal SYCL command-group submission.

* Throws synchronously with error code `invalid` if any of the `sycl::event`s
associated with `depEvents` came from the same graph that the graph node

> resulting from this command-group submission is associated with.

Reviewer: nitpick - monospace isn't rendering nicely (same for `handler::ext_oneapi_barrier(const std::vector<dynamic_event>& waitList)` description):
![image](https://github.com/reble/llvm/assets/2334151/6bca2b9f-2285-4f37-97aa-dac56cb37aff)

|===

=== Dynamic Events [[dynamic-events]]

[source,c++]
----
namespace ext::oneapi::experimental {
  class dynamic_event {
    dynamic_event();
    dynamic_event(const event& syclEvent);

>     void update(const event& syclEvent);

Reviewer: Noting a point that came up offline, that using events returned from host-tasks here might be difficult to implement (as not backed by a native event), so possibly forbid that until we can work out how to implement it.

Author: I've added error wording for using host_task events now.

  CGH.depends_on(DynEvent);
  CGH.parallel_for(...);
});
----
Dynamic events created with an event from a `command_graph` cannot then be
associated with other nodes in that same graph as this could be used change the
topology of the graph. Attempting to call `handler::depends_on()` with such a
`dynamic_event` in that situation will result in an error.

Dynamic events can be created with no event but must be updated with a valid
event before any executable graph which depends on that event is executed.
Failing to do so will result in an error.

The `dynamic_event` class provides the {crs}[common reference semantics].

Table {counter: tableNumber}. Member functions of the `dynamic_event` class.
[cols="2a,a"]
|===
|Member function|Description

|
[source,c++]
----
dynamic_event();
----

| Constructs a default `dynamic_event` which is not associated with any SYCL event.

|
[source,c++]
----
dynamic_event(const event& syclEvent);
----

| Constructs a `dynamic_event` which is associated with the SYCL event `syclEvent`.

Parameters:

* `syclEvent` - The SYCL event to associate this `dynamic_event` with.

|
[source,c++]
----
void update(const event& syclEvent);
----

> | Updates the SYCL event associated with `dynamic_event`. This update will be

Reviewer 7: I found this description a bit hard to understand on a first read through . I think I struggled with exactly what "`dynamic_event`" and "associated modifiable graph nodes" meant.

```
Updates the SYCL event associated with `dynamic_event`
```
I think it would make more sense to say "Updates the SYCL event associated with this object" or "Updates the SYCL event associated with this dynamic event".

```
This update will be reflected immediately in the associated modifiable graph nodes. An executable
graph can then be updated to reflect these new event dependencies using
<<executable-graph-update, Executable Graph Update>>.
```

Maybe it would be easier to say something along the lines: "This update will take immediate effect for nodes that are part of a modifiable graph. Executable graphs can be updated to reflect these new event dependencies using <<executable-graph-update, Executable Graph Update>>."

Author: > Maybe it would be easier to say something along the lines: "This update will take immediate effect for nodes that are part of a modifiable graph. 

I agree this wording is quite convoluted but I feel like this suggestion implies that nodes can optionally *not* be part of any modifiable graph which isn't the case. I think what I'mt rying to convey is that:

- Nodes are essentially the elements of the graph container/structure, and so updating them reflects those changes immediately because they *are* the modifiable graph state.
- Executable graphs are normally fixed, and so you provide the equivalent node from the modifiable graph with which to update the state of the executable graph based on that modifiable node's state.

reflected immediately in the associated modifiable graph nodes. An executable
graph can then be updated to reflect these new event dependencies using
<<executable-graph-update, Executable Graph Update>>.

Parameters:

* `syclEvent` - The new SYCL event to update this `dynamic_event` with.

Exceptions:

* Throws synchronously with error code `invalid` if `syclEvent` is an event
obtained from a queue submission which was recorded to the same graph any of the

> `node`s associated with this `dynamic_event` are from.

Reviewer: nitpick - but this line doesn't render nicely:
![image](https://github.com/reble/llvm/assets/2334151/3cebaa31-9689-487e-b8c0-133c60782b0c)

associated with the graph node resulting from this command-group submission is
different from the one with which the dynamic_parameter was created.

|
[source,c++]
----
void handler::ext_oneapi_barrier(const std::vector<dynamic_event>& waitList);
----

|Overload of `ext_oneapi_barrier` that takes a list of dynamic events to wait
on.

This function has the same semantics as `ext_oneapi_barrier(const
std::vector<event>&)`.

Parameters:

* `waitList` - List of dynamic event dependencies for this barrier.

Exceptions:

* Throws synchronously with error code `invalid` if this function is called from
a normal SYCL command-group submission.

* Throws synchronously with error code `invalid` if any of the `sycl::event`s
associated with `waitList` came from the same graph that the graph node
resulting from this command-group submission is associated with.

|
[source,c++]
----
void handler::depends_on(dynamic_event depEvent);
----

|Overload of `depends_on()` which takes a dynamic event dependency.

This function has the same semantics as `depends_on(event)`.

Parameters:

* `depEvent` - Dynamic event dependency.

> Exceptions:

Reviewer: Have we considered whether it should be an error if the `dynamic_event` isn't backed by a SYCL event at this point?

> Exceptions:

Reviewer: Did we consider whether it should be an error if the dynamic event isn't backed by a SYCL event at this point? This affects the UR design in that otherwise we have to provide a way to say that although a command has an empty event wait-list now, it can be updated in future. (not that this is a big deal, just whether it's a use case we expect).

recording mode. Features not listed below behave the same in recording mode as
they do in non-recording mode.

==== Event Limitations

For queue submissions that are being recorded to a modifiable `command_graph`,
the only events that can be used as parameters to `handler::depends_on()`, or
as dependent events for queue shortcuts like `queue::parallel_for()`, are events
that have been returned from queue submissions recorded to the same modifiable
`command_graph`.

> Other limitations on the events returned from a submission to a queue in the

Reviewer 3: I agree that the limitation in the previous paragraph can be removed.  However, don't we still have all these other limitations?

Author: Good point I was a bit hasty in removing those, thanks.

will depend on all the existing graph leaf nodes, not only the leaf nodes

> that were added from the queue being recorded.

Reviewer 3: Is it really the case still that barriers cannot be used when creating a graph via the explicit API?  It seems weird that we allow them when they use `dynamic_event` but not when they take a regular `event`.

Author: You are correct I think this restriction can be removed now as part of this PR.

- Creating external event dependencies for a graph, either
<<external-event-dependencies, one-time>> or <<dynamic-events, updatable between
executions>>.

==== Node-Level Execution Events [[node-execution-events]]

Events representing the completion of an individual node inside a given
executable graph can be obtained using
`command_graph<graph_state::executable>::get_event(node)`. These events can then
be waited on or used as dependencies for eager SYCL operations outside of
graphs. These events may be useful for operations which may be infrequent and
depend only on some intermediate results of work being done in the graph.

[source, c++]
....
sycl::event ExecutionEvent = ExecGraph.get_event(SomeNode);

Queue.submit((sycl::handler& CGH)
  {
    CGH.depends_on(ExecutionEvent);
    CGH.parallel_for(...);
  });

// The above operation will only execute once SomeNode has finished executing
// inside ExecGraph
Queue.ext_oneapi_graph(ExecGraph);
....

These events represent only the most recent execution of a given executable
graph. If an application executes the same graph multiple times before
scheduling work or performing a host-side wait on the event then executions of

> the node in a previous execution other than the most recent one may be missed.

Reviewer 2: for the long sentence, does this talk about the code like below?
```
sycl::event ExecutionEvent = ExecGraph.get_event(SomeNode);

for (int i = 0; i < 1000; ++i)  // 1000 is just one value for example
    Queue.ext_oneapi_graph(ExecGraph);

Queue.submit((sycl::handler& CGH)
{
    CGH.depends_on(ExecutionEvent);
    CGH.parallel_for(...);
});
```

what is "Queue.ext_oneapi_graph(ExecGraph); with i==998" missed?

Author: After having some internal discussions with regards to implementability of this aspect of the spec I've pushed some wording to clarify what these events represent, but I'll also address your specific example.

As it stands right now if multiple SYCL graph submissions are in-flight as in your example, it is undefined at the point where you call `CGH.depends_on(ExecutionEvent);` which exact execution it will depend on, as it depends which iteration is currently executing on device.

To depend on a specific execution the application must enforce ordering such that multiple enqueues are not in-flight, like so (example I just added taken from the spec):

```c++
sycl::event NodeExecutionEvent;
sycl::event GraphCompletionEvent;

for (size_t i = 0; i < GraphIterations; i++){
  // Obtain the node execution event for the graph
  NodeExecutionEvent = ExecGraph.get_event(SomeNode);

  // Enqueue the graph for execution
  sycl::event GraphCompletionEvent = Queue.ext_oneapi_graph(ExecGraph);
  // Use the event, to wait on the host and perform some intermediate
  // host-work once that node has completed
  ExecutionEvent.wait();
  DoSomethingOnHost();
  // Wait on the graph finishing to ensure it is complete before the next use of
  // a node-level execution event
  GraphCompletionEvent.wait_and_throw();
}
```

This is generally in keeping with how the backends handle this but slightly more restrictive because SYCL events do not function like CUDA/L0 events and the user does not have actual visibility of device execution because this is obfuscated by the SYCL runtime.

depend only on some intermediate results of work being done in the graph.

[source, c++]
....
sycl::event ExecutionEvent = ExecGraph.get_event(SomeNode);

Queue.submit((sycl::handler& CGH)
  {
    CGH.depends_on(ExecutionEvent);
    CGH.parallel_for(...);
  });

// The above operation will only execute once SomeNode has finished executing
// inside ExecGraph
Queue.ext_oneapi_graph(ExecGraph);
....

These events represent only the most recent execution of a given executable
graph. If an application executes the same graph multiple times before
scheduling work or performing a host-side wait on the event then executions of
the node in a previous execution other than the most recent one may be missed.
Applications requiring this should take care to schedule eager operations/waits
between each graph execution, or include these operations as nodes in the graph
if they are to be performed for every graph execution.

These events cannot be used to define dependencies between graphs. These should
instead be defined as described in <<inter-graph-dependencies, this section>>.

==== Adding External Event Dependencies To Graphs [[external-event-dependencies]]

> <<event-terminology, Regular SYCL events>> can be passed as dependencies to

Reviewer 2: Regular SYCL events include:
1) normal submissions to a SYCL queue
2) events returned from submitting an executable command_graph for execution
3) events obtained via command_graph<graph_state::executable>::get_event()

as for item 3, `These events can then be waited on or used as dependencies for eager SYCL operations **outside** of graphs`, so can item 3 be used as `dependencies to graph nodes`?

Author: No this is explicitly disallowed in the spec, dependencies between graph nodes should be established using graph-limited events.

Reviewer 2: If the answer is no, I think we can not just use "Regular SYCL events" here since not all the three parts are all allowed.

Author: That's fair, I've updated the wording here to refer to events from eager SYCL submissions only.

dynParamAccessor.update(bufferB.get_access());
----

=== Dynamic Event Update

> Example which shows create a dependency between two graphs using dynamic events

Reviewer 3: ```suggestion
Example which shows creation of a dependency between two graphs using dynamic events
```

or maybe "shows how to create"

queue myQueue;
auto myContext = myQueue.get_context();
auto myDevice = myQueue.get_device();

// Create the first graph
sycl_ext::command_graph graphA(myContext, myDevice);

// Add a node to graphA
sycl_ext::node nodeA = graphA.add(...);

sycl_ext::command_graph<sycl_ext::graph_state::executable> execGraphA = 
  graphA.finalize();

// Get an event for the next execution of nodeA in execGraphA
event eventA = execGraph.get_event(nodeA);

// Create the second graph
sycl_ext::command_graph graphB(myContext, myDevice);

// Create the dynamic event which will wrap eventA and use it as a dependency
// for a node in graphB
sycl_ext::dynamic_event dynEvent{eventA};

auto nodeB = graphB.add([&](handler& CGH){
  CGH.depends_on(dynEvent);
  CGH.parallel_for(...);
});

auto execGraphB = graphB.finalize({sycl_ext::property::graph::updatable});

// Submit both graphs for execution
myQueue.submit(execGraphA);
myQueue.submit(execGraphB);
myQueue.wait_and_throw();

// Update the dynamic event with a new event frome execGraphA to reflect the
// next execution
dynEvent.update(execGraphA.get_event(nodeA));
// Update execGraphB with the affected node to reflect those changes

> execGraphB.update(nodeB);

Reviewer 4: How optimized the graph update is? I don't have latency numbers at hand, but in CUDA, we try to avoid updating the graph when not absolutely needed, so I am worried that having to update it every step might negate most of the performance benefits of more streamlined submission.

Author: I don't have hard numbers, and graph update is a fairly new feature for us so there is likely room for improvement performance wise but I imagine there will always be some amount of overhead compared to not doing an update. I can see how that is a concern for your use case though.

One of the primary drivers for this design is a different use-case where the user wants the ability to wait for an event in the middle of the graph to be able to do some separate operations before the graph has fully finished executing. In that scenario it is important for us to be clear about what exactly each `sycl::event` represents, So it made sense to be more explicit about each event representing a single execution, which does require a graph update as you say.

We had previously looked at designs where events would be reset automatically but it is difficult to reason about that as a user for this primary use-case. I do think that makes more sense for your stated use-case however, since the exact semantics of how these events are handled doesn't matter to the user as I see it, but it does present a challenge from an API design point of view in SYCL trying to create something that makes sense for both use cases and the wider SYCL specification.

Reviewer 4: > I can see how that is a concern for your use case though.

I would argue that it's a concern for many use-cases. Graphs are primarily used for latency-sensitive workloads, right? And here we mandate that if we want to use dynamic events as a dependency, we must update the graph every time it's submitted.

> One of the primary drivers for this design is a different use-case where the user wants the ability to wait for an event in the middle of the graph to be able to do some separate operations before the graph has fully finished executing. In that scenario it is important for us to be clear about what exactly each sycl::event represents, So it made sense to be more explicit about each event representing a single execution

I do like the design in this PR more than the auto-resetting events / semaphore approach. Being explicit about which instance of the graph we are referring to is quite helpful in reasoning about the code (even if it is a bit more verbose). The question is, can it be efficient.

In particular, perhaps there is a way to leverage the fact that native CUDA events *are* auto-resetting, and make `dynamic_event` somehow smart enough to make the graph update a no-op when the same CUDA event can be reused?

Author: > In particular, perhaps there is a way to leverage the fact that native CUDA events are auto-resetting, and make dynamic_event somehow smart enough to make the graph update a no-op when the same CUDA event can be reused?

I think it is probably possible, but as I mentioned in another comment I think there's quite a high implementation burden for that. Since we both have to map to multiple, quite different APIs on the UR level, and both the graph runtime and UR implementations are very much not set up to support that kind of information sharing between graphs/command-buffers.

I think because of that burden if we were going to do that it would need to be done as a separate piece of work, so I would be in favour of leaving this design as is for now. It should be easy to update the spec to a new behaviour in the future, if we can commit to the implementation in the future.

Reviewer 2: > Graphs are primarily used for latency-sensitive workloads, right? And here we mandate that if we want to use dynamic events as a dependency, we must update the graph every time it's submitted.

my comment from user perspective, we always like the static graph without change, but if we are unable to keep the graph always the same (due to use case), we would have to use the dynamic (parameters or events) to update the graph, such overhead is likely to be less than recording/creating the graph again.

Author: > I would argue that it's a concern for many use-cases. Graphs are primarily used for latency-sensitive workloads, right? And here we mandate that if we want to use dynamic events as a dependency, we must update the graph every time it's submitted.

I was referring to your use-case in contrast to the use-case example that I gave which may not use these kind of dependencies between graphs, and thus would not need to use update for their use-case, rather than trying to imply that your use-case is the only one with such concerns.

I agree that this is a concern for most and is the primary driver behind using graphs. Ideally having to update, especially a small number of nodes/events, will not outweigh the gains in reduced host latency from using graphs so there would still be a net performance increase. The effectiveness of graphs as a whole is workload dependent though so this may not be true for all cases.

// Both graphs can now be executed again with updated dependencies between them.
myQueue.submit(execGraphA);
myQueue.submit(execGraphB);
myQueue.wait_and_throw();
----

> 

Reviewer 4: If we're using the record-replay mode, is the proposed solution to do the following?

```cpp
// Graph `g1` is recording queue `q1`....
sycl::event e1  = q1.submit(....); // Submit something
auto node1 = sycl::node::get_node_from_event(e1); // Get persistent node handle
g1.end_recording();
eg1 = g1.finalize(); // Get executable graph
sycl::dynamic_event de{e1};

// Graph `g2` is recording `q2`; we want it to depend on `e1`.
sycl::event e2 = q2.ext_oneapi_submit_barrier({de});
auto node2 =  sycl::node::get_node_from_event(e2);
g2.end_recording();
eg2 = g2.finalize();

// Every time `eg1` is submitted, do the following:
q1.ext_oneapi_graph(eg1);
dynEvent.update(eg1.get_event(node1));
eg2.update(node2);
```

Author: Yes that would be the intended way of using it in record & replay, although that raises a good point that I need to add wording about extending `ext_oneapi_submit_barrier` to allow passing `dynamic_event`s instead of normal events.

Reviewer 2: I have one question about `e1` used in `sycl::dynamic_event de{e1};`, it is generated within the graph recording stage (before g.finalize() is called).

while in the code below for the explicit API case, `eventA`, used in `sycl_ext::dynamic_event dynEvent{eventA};` is generated with `event eventA = execGraph.get_event(nodeA);` (after g.finalize() is called).

I'm wondering if `e1` and `eventA` have the same underlying implementation? (or they are two different concepts?)

Author: Ah I missed that in the example, it is not valid to use `e1` as a parameter to the dynamic event, instead it should be:

```c++
sycl::dynamic_event de{eg1.get_event(node1)};
```

I wouldn't say they have a different implementation - they are both SYCL events. However, events returned from recording submissions are not attached to backend events (since the backend structures for graphs are not defined until finalize has been called, and thus these events are not associated with the execution of anything) and so can only be used for defining intra-graph dependencies.

dynParamAccessor.update(bufferB.get_access());
----

=== Dynamic Event Update

Example which shows how to create a dependency between two graphs using dynamic
events which are then updated between executions.

[source,c++]
----
...

using namespace sycl;
namespace sycl_ext = sycl::ext::oneapi::experimental;

queue myQueue;
auto myContext = myQueue.get_context();
auto myDevice = myQueue.get_device();

// Create the first graph
sycl_ext::command_graph graphA(myContext, myDevice);

// Add a node to graphA
sycl_ext::node nodeA = graphA.add(...);

sycl_ext::command_graph<sycl_ext::graph_state::executable> execGraphA = 
  graphA.finalize();

// Get an event for the next execution of nodeA in execGraphA
event eventA = execGraph.get_event(nodeA);

// Create the second graph
sycl_ext::command_graph graphB(myContext, myDevice);

// Create the dynamic event which will wrap eventA and use it as a dependency
// for a node in graphB
sycl_ext::dynamic_event dynEvent{eventA};

auto nodeB = graphB.add([&](handler& CGH){

>   CGH.depends_on(dynEvent);

Reviewer: I don't think we've added an entry to our table of new `handler` methods an entry for `handler::depends_on(sycl_ext::dynamic_event)`.

I only see `depends_on` extended for `sycl_ext::property::node::depends_on`
